src/pages/Lending/Pool/components/Repay/RepayInput.tsx

```diff
@@ -130,10 +148,110 @@ export const RepayInput = ({
     setConfirmedQuote(lendingQuoteCloseData ?? null)
   }, [lendingQuoteCloseData, setConfirmedQuote])
 
+  const userAddress = useMemo(() => {
+    if (!repaymentAccountId) return ''
+
+    return fromAccountId(repaymentAccountId).account
+  }, [repaymentAccountId])
+
+  const { data: inboundAddressData, isLoading: isInboundAddressLoading } = useQuery({
+    ...reactQueries.thornode.inboundAddresses(),
+    staleTime: 60_000,
+    select: data => selectInboundAddressData(data, repaymentAsset?.assetId),
+    enabled: !!repaymentAsset?.assetId,
+  })
+
+  const { data: allowanceData, isLoading: isAllowanceDataLoading } = useAllowance({
+    assetId: repaymentAsset?.assetId,
+    spender: inboundAddressData?.router,
+    from: userAddress,
+  })
+
+  const isApprovalRequired = useMemo(() => {
+    if (!confirmedQuote) return false
+    if (!repaymentAsset) return false
+    if (!isToken(fromAssetId(repaymentAsset.assetId).assetReference)) return false
+
+    const supportedEvmChainIds = getSupportedEvmChainIds()
+    if (
+      !supportedEvmChainIds.includes(fromAssetId(repaymentAsset.assetId).chainId as KnownChainIds)
+    )
+      return false
+
+    const allowanceCryptoPrecision = fromBaseUnit(allowanceData ?? '0', repaymentAsset.precision)
+    return bnOrZero(confirmedQuote?.repaymentAmountCryptoPrecision).gt(allowanceCryptoPrecision)
+  }, [allowanceData, confirmedQuote, repaymentAsset])
+
   const mixpanel = getMixPanel()
+
+  const repaymentAccountNumberFilter = useMemo(
+    () => ({ accountId: repaymentAccountId }),
+    [repaymentAccountId],
+  )
+  const repaymentAccountNumber = useAppSelector(state =>
+    selectAccountNumberByAccountId(state, repaymentAccountNumberFilter),
+  )
+  const {
+    mutate,
+    isPending: isApprovalMutationPending,
+    isSuccess: isApprovalMutationSuccess,
+  } = useMutation({
+    ...reactQueries.mutations.approve({
+      assetId: repaymentAsset?.assetId,
+      spender: inboundAddressData?.router,
+      from: userAddress,
+      amount: toBaseUnit(
+        // Add 5% buffer to the repayment allowance to avoid asset rates fluctuations ending up in more asset needed to repay
+        bnOrZero(confirmedQuote?.repaymentAmountCryptoPrecision).times(1.05),
```

Comment:
for the time being, we can assume/hope this buffer is sufficient, but long term it seems like a good idea to stop polling quotes once the repay flow is initiated via approve or actual repay to keep the quote static and the values known.

The expiry could be used of course as a means of reset mechanism for a user who waited too long between approval -> repay in which they need to go back through the flow as a safeguard.

Reply:
Big 🧠  - much better than a buffer, and we do check for expiry at repay confirm step already!
9b55fddd86

- Approval needed

https://github.com/shapeshift/web/assets/17035424/ebd12da2-2ef7-4d1d-a262-d0928af385b7

- Approval *not* needed


https://github.com/shapeshift/web/assets/17035424/0dc6ab9c-8bbe-4f76-9275-c6bb37b99ced

Same issue here since I wasn't able to test with an actual Tx, there is a slight risk that we rerender -> `enabled` is true again -> we refetch the quote and then end up in the same situation.

To be confirmed in ~5 1/2 days when my repayment unlocks or if/when someone else is able to test it.